### PR TITLE
Improve efficiency of outer map operations for Roaring64Map

### PR DESCRIPTION
**Pull request by kosak**
Hello,

The purpose of this PR is to slightly improve the efficiency of `Roaring64Map` in C++.

As you know `Roaring64Map` contains an "outer" `std::map`, which maps the key's high bits to individual (32-bit) Roaring maps, which in turn handle the key's low bits. The problem is that many of the methods in `Roaring64Map` probe the outer map more times than necessary. This includes high-traffic items like `add()` and `contains()`, which probe twice when they only need to probe once.

I've changed a variety of methods to do fewer probes. While I was here I also took the liberty to make a few changes:
* In `flip()`, I added what I believe to be a missing call to `setCopyOnWrite()`
* In `operator &=`,  `operator -=`, and `operator ^=`, if the 32-bit Roaring entry becomes empty, I remove it from the outer map altogether, rather than leaving an empty Roaring in that slot. I believe this is preferable (?)
* In `shrinkToFit()` I removed an instance of what I believe is undefined behavior. In the line `roarings.erase(iter++)` I believe it is theoretically possible for the increment to happen after the call to `erase` has finished (and therefore `iter` would be an invalid iterator at that point and you ought not try to increment it).  EDIT: no, this can't happen with an overloaded operator, so, forget I mentioned it. (I would still prefer to keep my change because I feel it reads better).
* In `maximum()` I changed the code to use forward iterators, but traversing in the reverse direction, rather than reverse iterators. The rationale is that map's `reverse_iterator` is significantly slower than its `forward_iterator`, because in `reverse_iterator`, the `operator *` and `operator ->` methods are not constant time because they both do a decrement operation on a temporary internal tree iterator. I can provide more detail if you're interested.

The code passes the current test suite. I'm not sure if that's sufficient or if additional tests need to be written.

### CODE REVIEW

**Comment by lemire** [comment]
@SLieve  Would you review?

**Comment by SLieve** [comment]
Sure!

**Comment by kosak** [comment]
Looking over the code again, I need to retract what I said about undefined behavior and iter++. The issue I mentioned isn't relevant with an overloaded operator. Nevertheless, I still like my change to the code, as I think it reads better.

File: cpp/roaring64map.hh
Comment (SLieve) [inline cpp/roaring64map.hh:387]: Nit: Could we assign `result.first->second` to a variable to clarify what it is? Same elsewhere.

**Review by SLieve — CHANGES_REQUESTED** [review]
Changes look great, I especially like the clear comments.

The current test suite is fairly sparse. I think it would make sense to add separate unit tests at least for `&=, -=, |=, ^=`, and `rank` to `tests/cpp_unit.cpp`. Optionally also around copy on write functionality, though I understand that's much harder to test and would probably be better off as a separate PR.

**Comment by kosak** [comment]
Thanks for the replies! I'll get back to you in a little while with updates and unit tests.

File: cpp/roaring64map.hh
Comment (lemire) [reply cpp/roaring64map.hh]: I stress that claims related to performance ought to be backed by benchmark results.

**Comment by kosak** [comment]
Hi,

I've done a pretty deep dive on this code. The more I looked at it, the more stuff I felt I wanted to change. 

I realize this may be more than you want, and you may decide you want none, some, or all of these changes. Or you might want this large pull request broken up into smaller pieces, or you may want more unit tests.

In any case I thought I would show you what I have, and you can decide how you would like to proceed.

Here is a list of what I changed:
* various documentation and grammar nits
* Renaming variables/arguments as well as adding temporary variables for readability
* Unit tests: fixed some compiler warnings, changed some `assert()` to `assert_true()`
* Additional per-function behavioral changes since last time

The last item involved a lot more radical changes so I will go into more detail here to explain my rationale on a method-by-method basis:

`bitmapOf(std::initializer_list<uint64_t> list)`: a safe, more idiomatic way of passing a list of items that is less error-prone than C-style varargs

More modifications to: `operator&=`, `|=`, `^=`, `-=` : (simple) logic to do the right thing when the left and right hand sides are the same object.  as in `mymap  -= mymap`

Added `cardinality_nothrow()`: The existing `cardinality()` has a caveat that warns that it might throw an exception if the map is completely full, and that users who care should call `isFull()` first. I'm not sure I really believe in this scenario (2^64 is a lot of bits), but the user who finds themselves in that scenario will first have to call a (very slow) `isFull()` followed by a (very slow) cardinality(), which seems a little user-unfriendly. I felt it would be more user-friendly to provide a `cardinality_nothrow()` so that users can get their information in one call rather than two. As a happy side effect, both `cardinality()` and `isFull()` can now simply delegate to `cardinality_nothrow()` so the net amount of code is smaller.

I also find myself wishing that `cardinality()` was constant time rather than (painful to calculate) but I assume there are reasons it is the way it is.

`isStrictSubset()`: The existing `isStrictSubset()` warns that it might throw if the bitset is completely full, and like the above, users who might be in this scenario need to call `isFull()` first. This felt a little user-unfriendly to me, because it leaks an implementation detail that the user shouldn't have to care about in the first place. What I did was make a helper method  bool `isSubset(const Roaring64Map &other, bool requireStrict)`, that both `isSubset` and `isStrictSubset` can delegate to. This new method scans the map and determines strictness along the way. This new helper method is public but doesn't need to be.

`toUInt64Array`: overuse of functional style obfuscates a simple method (in my opinion!)

`flip`: A little background here. The standard style for other methods (`addRange` vs `addRangeClosed`, `removeRange` vs `removeRangeClosed`) is to provide a "half-open interval" version and a "closed interval" version, and the former would simply adjust the coordinates and call the latter. It appears in this case that the underlying 32-bit bitmap only has a "half-open range" version of flip(). So the 64 bit version was written to call that, but still uses the same coding recipe that the other closed versions do. I suspect this might cause some bugs to creep in but I'm not sure this actually caused any bugs. However I can point to one place where it comes to a wrong conclusion. As a counterexample, if you pass it the half-open interval [0xFFFFFFFF, 0x100000000), you would expect it to conclude that that interval falls on a single underlying bitmap and invoke that clause in the code. However, it does not come to that conclusion, because highBytes(0xFFFFFFFF) != highBytes(0x100000000) and so it fails its (incorrect) same-underlying test. Again I don't think that breaks anything but it's a sign that the code is a little incoherent. I felt it was better to provide two versions of `flip()`: `flip()` and `flipClosed()`. This allows the code to follow the same tried-and-true recipe as already established for `addRange()`/`addRangeClosed()` and `removeRange()`/`removeRangeClosed()`. Now if we were strictly following the recipe, we would have the problem that flipClosed() needs to call the corresponding 32-bit `flipClosed` (which doesn't exist for some reason). But that's fine, we can work around it with another coordinate adjustment and be on our way. A perhaps better alternative would be for the 32-bit code to actually provide both `flip()` and `flipClosed()`, but that's your decision. In any case the code follows the standard recipe now and is therefore hopefully less likely to be buggy.

`select`: Forcibly casts a uint64 pointer to a uint32 pointer, which is likely undefined behavior, and comes with the warning `// assuming little endian`, which I fear means that big-endian customers get corrupt data. Anyway, there's no need for this. A small rewrite to this function eliminates both issues.

`printf` and `toString`: Another overuse of the functional style makes what should be a simple method virtually impossible to read (in my opinion!). Instead I provide a standard C++ operator for streaming to an `ostream`, which is more idiomatic C++. And now, `printf` and `toString` can simply delegate to that.

`fastunion`: So, `operator|=` has a comment `See also the fastunion function to aggregate many bitmaps more quickly.` but then fastunion has the disclaimer `// not particularly fast` and it turns out that it just iterates over the sets and unions them one-by-one. So I rewrote `fastunion` to basically do a "group by" operation, grouping the sets by their highbits key, and then calling out to `roaring_bitmap_or_many` to combine each group. 

**UNIT TESTS**: I added unit tests for `operator&=`, `|=`, `-=`, and `^=`. Basically I generate two large sets, populate them sort-of randomly and then perform the operations on them. I couldn't really decide whether this should be in `cpp_unit.cpp` or `cpp_random_unit.cpp`, so I put them in the former for now. Also, while it was tempting to reuse the `make_random_bitset64` function, I felt I wanted more like a million values, not 100, and I wanted to distribute them over the two sets in an "interesting" way. Probably a little more work is needed here but I wanted your feedback first.

Also cleared out a few warnings. Also there are a bunch of `assert` calls in `toplevel_unit.c` which I'm pretty sure were meant to be `assert_true()`, since "assert" is a no-op in optimized builds and I assume you wanted these unit test assertions evaluated regardless of optimization setting.

This PR passes all existing unit tests.

Again, kindly let me know if you want less, more, split into multiple PRs, more unit tests, etc.

File: cpp/roaring64map.hh
Comment (kosak) [reply cpp/roaring64map.hh]: See my "mega-comment" below.

File: cpp/roaring64map.hh
Comment (kosak) [reply cpp/roaring64map.hh]: Done.

File: cpp/roaring64map.hh
Comment (kosak) [reply cpp/roaring64map.hh]: Fair enough. I've looked at the assembly code and I wrote a pretty thorough benchmark. The assembly code does show extra tree operations (as I expected) but my benchmark is all over the place: given a set of benchmark parameters sometimes one way is faster, sometimes the other way is faster, and sometimes they're more or less the same. At this point I assume that CPU caching/prediction/etc is papering over the inefficient assembly code I'm seeing. Given that I can't seem to write a slam-dunk proof of superiority, I'm going to retract that change for now. Let me know if you care to see the benchmark.

**Comment by SLieve** [comment]
To make the non-functional changes easier to review, I suggest moving them to a separate PR. Of course the line is a bit blurry here, but I'll leave it up to you to make the distinction.

**Comment by madscientist** [comment]
Please do not add a dependency on the C++ stream library into CRoaring.  If you want something like that at least move it to a separate/separable header/source so that it can be omitted on systems where stream is not available/wanted/present.

**Comment by madscientist** [comment]
I have no problem with select being rewritten but note that it is not undefined behavior to cast an unsigned 64bit value to 32bits: the standard defines this (to truncate the value).

**Comment by kosak** [comment]
> I have no problem with select being rewritten but note that it is not undefined behavior to cast an unsigned 64bit value to 32bits: the standard defines this (to truncate the value).

Of course. But the code In question wasn't casting a value, it was casting a pointer (from `uint64_t*` to `uint32_t*`):

```
// assuming little endian
return map_entry.second.select((uint32_t)rnk, ((uint32_t *)element));
```

**Comment by kosak** [comment]
@SLieve OK. I could go a little more fine-grain if you want. The changes are pretty independent. I could do separate PR's for
* &=, ^=, |=, -=  (they all look the same)
* improving the print routines (without involving the C++ stream library, per @madscientist )
* comment-only changes
* cleanups in existing unit tests (like the change assert -> assert_true and uninitialized variable warnings)
* and then maybe one PR per major functional change, like
** a PR for fastunion
** a PR for cardinality_nothrow
** a PR for flip

Or is that too *many* PRs?

Just let me know. I'll get back to this in a couple of days.

**Comment by SLieve** [comment]
@kosak that works for me!

**Comment by lemire** [comment]
This is very good work and we want to merge this. Unfortunately, it seems we got off-track.

@kosak Comments?

**Comment by kosak** [comment]
Hi, thanks for your kind words and I'm sorry for the delay. I said I would come back to this in "a few days" but then life got in the way :-). Anyway, as promised I have broken my change up into smaller pieces to make it easier to review. I have recycled this PR to focus on just this first step. (I could also move this to a new PR if you want).

Here is the first piece, which deals only with the overloaded set operations (`&=`,`|=`,`-=`,`^=`)
In addition to the usual things, I would ask the reviewers to also consider the following issues:
1. I am adding an "auto-cleanup" behavior that wasn't there before. What this means is, if the code detects that it has resulted in an empty inner Roaring, it will delete it from the outer C++ map rather than leaving an empty "placeholder" in that outer slot. Personally I think it's preferable to do it this way.
2. I added tests for these operators but couldn't decide where they belong (cpp_unit vs cpp_random_unit). cpp_unit seems to be for the "main" tests and cpp_random_unit is for additional random challenges. So I *think* these tests should be in cpp_unit even though they do use a random strategy. Furthermore I used the C++ random number generator rather than rand() because I wanted a local decent-quality random number generator with deterministic behavior. Also note that these tests don't need to thoroughly test the underlying 32-bit Roaring (that is done elsewhere), but rather they need to focus on the outer 64-bit bucket management.
3. I also added lots of comments and used the terminology "XOR" rather than "symmetric union" which I think is more common.

**Comment by SLieve** [comment]
Great, thanks for splitting it up! I'll try to review it soon.

**Review by SLieve — COMMENTED** [review]
Looks great, very clear improvements!

File: cpp/roaring64map.hh
Comment (kosak) [reply cpp/roaring64map.hh:405]: In principle I would agree with you. In practice this performs rather poorly.

The `std::map::lower_bound` member function knows it is working with a sorted tree data structure and is able to navigate to its target in `lg(N)` time.

On the other hand, the `std::lower_bound` free function is implemented by code similar to that shown here https://en.cppreference.com/w/cpp/algorithm/lower_bound

This code basically uses a binary search algorithm to find the right spot and also runs in `lg(N)` *operations*. But for the method to run in  `lg(N)` *time*, the underlying iterator subtraction and advance-by-k operations need to be constant time operations. If you are looking at the reference code linked above, the iterator ops I'm talking about happen inside `std::distance` and `std::advance`. For a container like `std::vector`, that constant-time assumption holds, and `std::lower_bound` works as expected.

The problem for us here is that both those operations are *linear time* for iterators that come from `std::map`. For example `std::distance(map.begin(), map.end())` will traverse the whole map and therefore costs O(`map.size()`). Likewise `std::advance(iter, k)` costs O(k) for a map iterator. Ultimately this means that the `std::lower_bound` free function is linear time when applied to a `std::map` and once things are going that slowly, you'll be better off just iterating over the whole map yourself.

If I may pontificate for a moment, I would say there's no *algorithmic* reason why the library maintainers can't provide a member function for `std::map` that would work like you suggest. And also there's no *algorithmic* reason why `std::map` couldn't use a slightly fancier underlying tree structure that would support faster distance and advance operations so std::lower_bound could run faster. But in the current reality, we don't have either of the above.

File: tests/cpp_unit.cpp
Comment (kosak) [reply tests/cpp_unit.cpp:874]: Kindly allow me to defend the approach a little bit, and then you can decide ultimately what to do about it. The `make_two_big_roaring64_maps` generator populates the bitsets with deterministic pseudorandom  keys which fall into four outer buckets:

A: keys that are likely only in set 1
B: keys that are likely only in set 2
C: keys that are in both sets
D: keys that are likely in the same outer bucket but different inner positions

The value of the keys don't *really* matter that much, and the fact that they're spread all over the `uint64_t` space is basically fine in my opinion. I'm not expecting them to ever collide (but it doesn't matter if they do). It's much more interesting to observe which of cases A-D they fall into.

I say "likely" because due to random effects we might get lucky or unlucky with duplicate keys but this doesn't really matter. I don't dispute your math in the above comment, but hitting the edge case neither causes anything interesting to happen nor breaks anything. Put another way, in case D our *intention* is to create two different keys that happen to belong to the same outer bucket but in different inner positions. Every once in a great while, if we happen to hit your edge case, in fact those keys end up in *different* outer buckets and different positions. So every once in a while we thought we were making a case D but we ended up making a case A plus case B. But it's all random anyway, so we don't care and getting lucky or unlucky doesn't break anything. There are lots of other ways to get lucky or unlucky (we might have a case A followed by the exact same key for case B and so we've "really" created a case C but again it's sort of like, that's the luck of the draw, no harm no foul).

The reason I implemented pattern A-D above is that it creates interesting cases for union, intersection, difference, and XOR. For example, given a million keys we would expect:
set 1 to have approximately 750K keys (250K each from cases A, C, and D)
set 2 to have approximately 750K keys (250K each from cases B, C, and D)
`(set1 & set2)` to have approximately 250K keys (from case C)
`(set1 | set2)` to have approximately 1.25M keys (250K each from cases A-C and 500K from case D)
`(set1 - set2)` to have approximately 500K keys (250K each from case A and case D)
`(set1 ^ set2)` to have approximately 1M keys (250K each from cases A and B, and 500K from case D)

When we add print statements that print these cardinalities, we see numbers that closely match these predictions:
set 1 size 750787
set 2 size 749667
intersection 250230
union 1250224
difference 500557
XOR 999994

So we have both the exact comparison from `does_std_set_match_roaring()` and also the aggregate statistical expectation which provides a little comfort that things are working as intended. Note I do not have print statements in the above code but I could print statements and/or assertions that say e.g. that union should result in exactly 1250224 elements.

Anyway, that's the rationale. I'm happy to proceed however you think makes sense. I do like your suggestion of iterating/refining the unit tests in a separate PR.

File: tests/cpp_unit.cpp
Comment (kosak) [reply tests/cpp_unit.cpp]: Please see the above regarding my rationale and then maybe you can decide how it's best to proceed.

File: cpp/roaring64map.hh
Comment (kosak) [reply cpp/roaring64map.hh:405]: Ah, I missed this comment in cppreference, which supports my point more succinctly

Complexity
The number of comparisons performed is logarithmic in the distance between first and last (At most log
2(last - first) + O(1) comparisons). However, for non-[LegacyRandomAccessIterators](https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator), the number of iterator increments is linear. Notably, [std::map](https://en.cppreference.com/w/cpp/container/map), [std::multimap](https://en.cppreference.com/w/cpp/container/multimap), [std::set](https://en.cppreference.com/w/cpp/container/set), and [std::multiset](https://en.cppreference.com/w/cpp/container/multiset) iterators are not random access, and so their member lower_bound functions should be preferred.

Comment (SLieve) [reply cpp/roaring64map.hh:405]: I missed that, thanks for the explanation and reference.

File: tests/cpp_unit.cpp
Comment (SLieve) [reply tests/cpp_unit.cpp:874]: Your rationale makes sense to me. I had actually misinterpreted the comment in the test to mean two adjacent outer positions rather than the same outer position. I think we should still add tests focused on specific scenarios, but let's leave that for later.

**Review by SLieve — APPROVED** [review]
Thanks for the discussion, this looks great and I think we're good to go

**Comment by lemire** [comment]
Running tests. If they successful, we might merge.

**Comment by lemire** [comment]
Merged.

**Pull request merged**
